src/vision/deliverables/borrowed_data_and_cancellation.md

```diff
@@ -0,0 +1,117 @@
+
+Under systems like [Swift's proposed structured concurrency model](https://github.com/apple/swift-evolution/blob/main/proposals/0304-structured-concurrency.md), or with APIs like [.NET's CancellationToken](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=net-5.0), cancellation is "voluntary". What this means is that when a task is cancelled, a flag is set; the task can query this flag but is not otherwise affected. Under structured concurrency systems, this flag is propagated to all chidren (and transitively to their children).
+
+[preemption]: https://tokio.rs/blog/2020-04-preemption
+
+Voluntary cancellation is a requirement for [scoped access](./scoped.md). If there are parallel tasks executing within a scope, and the scope itself is canceled, those parallel tasks must be joined and halted before the memory for the scope can be freed.
+
+One downside of such a system is that cancellation _may not_ take effect. We can make it more likely to work by integrating the cancellation flag into the standard library methods, similar to how tokio encourages ["voluntary preemption"][preemption]. This means that file reads and things will start to report errors (`Err(TaskCanceled)`) once the task has been canceled. This has the advantage that it exercises existing error paths and permits recovery.
+
+### Cancellation and `select`
+
+The `select` macro chooses from N futures and returns the first one that matches. Today, the others are immediately canceled. This behavior doesn't play especially well with voluntary cancellation. There are a few options here:
+
+- We could make `select` signal cancellation for each of the things it is selecting over and then wait for them to finish.
+- We could also make `select` continue to take `Future` (not `Async`) values, which effectively makes `Future` a "cancel-safe" trait (or perhaps we introduce a `CancelSafe` marker trait that extends `Async`).
+  - This would mean that typical `async fn` could not be given to select, though we might allow people to mark `async fn` as "cancel-safe", in which case they would implement `Future`. They would also not have access to ordinary async fn, though.
+    - Effectively, the current `Future` trait becomes the "cancel-safe" form of `Async`. This is a bit odd, since it has other distinctions, like using `Pin`, so it might be preferable to use a 'marker trait'.
+  - Of course, users could spawn a task that calls the function and give the handle to select.
+
+## Frequently asked questions
+
+### Could there be a convenient way to access the current scope?
+
+If we wanted to integrate the idea of scopes more deeply, we could have some way to get access to the current scope and reference its lifetime. Let's imagine we added a keyword `scope`, and we said that its lifetime is `'scope`. One would then be able to do something like the following:
+
```

Comment:
I'm just going to delete that. After further exploration I didn't find much reason to go this path, and I don't think it would fit that well with Rust. Also, to be honest, I'm not entirely sure I had in mind.